### PR TITLE
CareerQuiz Back End Logic

### PR DESCRIPTION
**Pull request by LuckyHariharan**
EDIT: This PR is now ready for review! Demo: https://share.vidyard.com/watch/Z6P5m67K8bb9fGFvQCMMYn?
@vithushan19 
The implementation of the backend logic uses:
- Insert Mutation for the StartQuiz component, to save name and email, and store id in state for future update mutations
- Update education mutation for the Education component, to save education, institution, degree and experience
- Update mutation for result and all other selections (industries, skills, tasks), 
- handleQuizMutations implemented in the handleNextClick, which checks the current stage state (START, EDUCATION, else) and uses the appropriate graphql mutation
- (Refactored Education tests to reflect these changes)

~~In this PR, I've setup the Backend Logic for the Career-Quiz so that:~~ 

~~- Starting the quiz immediately saves the users name and email using an upsert mutation~~ 
~~- Reaching the results stage of the quiz updates the user selections using the same upsert~~ 
~~- Navigating back to start and re-entering a name or email inserts a new row~~ 

~~Smaller changes:~~ 
~~- Fixed handler to allow maxselection functionality~~ 
~~- Refactored Education page unit tests to work properly, since the education stage has changed in this PR~~

### CODE REVIEW

File: components/resources/quizzes/shared/StartQuiz.tsx [hunk 1]
```diff
@@ -1,18 +1,38 @@
...
+const StartQuiz = ({
+  onNextClick,
+  body,
+  title,
+  userInput,
+  setUserInput,
+  handleUserInputMutations,
+}: StartQuizProps) => {
+  const handleInputChange = (event: React.ChangeEvent<HTMLInputElement>) => {
+    const { name, value } = event.target;
```
Comment (LuckyHariharan) [inline components/resources/quizzes/shared/StartQuiz.tsx:29]: save the users name and email in state as soon as they enter it

File: components/resources/quizzes/careerQuiz/career-quiz.tsx [hunk 2]
```diff
@@ -87,21 +127,58 @@ const CareerQuiz = () => {
...
-          ? { ...questionOption, isSelected: true }
-          : questionOption
-      ),
-    }));
-
-    const updatedQuizViewState = {
-      ...quizViewState,
-      questions: selectedQuizOption,
-    };
+    const currentQuestion =
```
Comment (LuckyHariharan) [inline components/resources/quizzes/careerQuiz/career-quiz.tsx:130]: updated handler to implement max selections

File: components/resources/quizzes/careerQuiz/career-quiz.tsx [hunk 3]
```diff
@@ -20,48 +33,75 @@ export enum Stage {
...
+  };
+  const [userInput, setUserInput] = useState({
+    name: "",
+    email: "",
+  });
+  const handleQuizResponseMutations = (
+    quizViewState: QuizViewState,
+    result: string
+  ) => {
+    const finalResponseObject = [
```
Comment (LuckyHariharan) [inline components/resources/quizzes/careerQuiz/career-quiz.tsx:77]: final response object used to pass user selections to the upsert mutation

File: components/resources/quizzes/careerQuiz/career-quiz.tsx [hunk 4]
```diff
@@ -1,17 +1,30 @@
...
-} from "../shared/SkillSelections";
+import SkillSelections from "../shared/SkillSelections";
 import StartQuiz from "../shared/StartQuiz";
+import { QuizOptionViewState, QuizViewState } from "../shared/types";
 import CareerResults from "./CareerResults";
-import EduBackground from "./EduBackground";
-
+import EduBackground, { EducationLevel } from "./EduBackground";
+const initializeQuizViewState = {
+  title: quizData.title,
```
Comment (LuckyHariharan) [inline components/resources/quizzes/careerQuiz/career-quiz.tsx:14]: initialize quiz to remove unnecessary useEffect

File: components/resources/quizzes/careerQuiz/career-quiz.tsx [hunk 5]
```diff
@@ -87,21 +124,58 @@ const CareerQuiz = () => {
...
-    };
+    const currentQuestion =
+      quizViewState.questions[quizViewState.currentQuestion];
+    const selectedOptions = currentQuestion.options.filter(
+      (option) => option.isSelected
+    );
+    const maxSelection =
+      quizData.questions[quizViewState.currentQuestion].maxSelections;
+
+    if (maxSelection === 1) {
```
Comment (iamnotreddy) [inline components/resources/quizzes/careerQuiz/career-quiz.tsx:135]: i think this logic is too long and a little over-complicated.. i played around with some code and got a simpler approach very close to working

check my suggestion and lmk if i'm overlooking something: 
- move the optional maxSelections prop to the QuizQuestionViewState type
- change how you're initializing the quiz view state so it accounts for this type change
- now you can access maxSelections in the skillSelections prop
- change the answer selection from a div to a button so you have access to the disabled prop
- store the number of questions selected in a local state and increment / decrement inside your handle click
```
const handleClick = (option: QuizOptionViewState) => {
    handleOptionClick(option);
    option.isSelected
      ? setNumOptionsSelected((prev) => prev - 1)
      : setNumOptionsSelected((prev) => prev + 1);
  };
```
- disable the button if num questions selected greater than max selections  
```
disabled={numOptionsSelected >= maxSelections && !option.isSelected}
```

only thing i didn't fully figure out was resetting the state when you switch to a different question

File: components/resources/quizzes/careerQuiz/career-quiz.tsx [hunk 6]
```diff
@@ -20,48 +33,72 @@ export enum Stage {
...
-          options: question.options,
-        };
-      }),
-      currentQuestion: 0,
-      progress: 0,
-    };
+  const [quizViewState, setQuizViewState] = useState<QuizViewState>(
+    initializeQuizViewState
+  );
+  const [saveUserPreferences] = useMutation(UPSERT_CAREER_QUIZ_RESPONSE);
```
Comment (iamnotreddy) [inline components/resources/quizzes/careerQuiz/career-quiz.tsx:49]: check the comment I left on Jimi's PR but we only need the below mutation function and we can call it saveQuizResponse

you'll still have two handlers though

**Review by iamnotreddy — CHANGES_REQUESTED** [review]
nice work! i think we can clean up the max selection logic, check out my approach below in the comments

One UI comment:
- let's just have it say "Select your highest level of education"
Screen Shot 2023-04-25 at 10 32 08 AM

File: components/resources/quizzes/shared/BluePrint.tsx [hunk 7]
```diff
@@ -3,12 +3,24 @@ import { Button } from "../../../ui/Button";
...
 };
 
-const BluePrint = ({ onNextClick, onBackClick }: BluePrintProps) => {
+const BluePrint = ({
+  onNextClick,
+  onBackClick,
+  handleQuizResponseMutations,
+  quizViewState,
+}: BluePrintProps) => {
+  const result = "Software Engineer";
```
Comment (vithushan19) [inline components/resources/quizzes/shared/BluePrint.tsx:23]: we don't career quiz specific logic in the shared components

File: components/resources/quizzes/shared/BluePrint.tsx [hunk 8]
```diff
@@ -40,7 +52,10 @@ const BluePrint = ({ onNextClick, onBackClick }: BluePrintProps) => {
       <div className="grid py-6 place-items-center">
         <Button
           label="View results"
-          onClick={onNextClick}
+          onClick={() => {
+            onNextClick();
+            handleQuizResponseMutations(quizViewState, result);
```
Comment (vithushan19) [inline components/resources/quizzes/shared/BluePrint.tsx:57]: can we just make this function part of the `onNextClick` function that's passed in from the parent?

**Review by iamnotreddy — APPROVED** [review]
as we discussed, im good with this if vithushan is

**Pull request merged**